### ISSUE TITLE
Horovd + AMP Contrastive Learning Imprecision

### ISSUE THREAD

**Issue by Zasder3** [horovod/horovod#3350]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet): PyTorch
2. Framework version: 1.10.1 (also tested on 1.7.0)
3. Horovod version: 0.23.0
4. MPI version: 4.1.1
5. CUDA version: 11.3
6. NCCL version: 2.8.3
7. Python version: 3.6.13
10. OS and version: Rocky Linux Release 8.5
11. GCC version: 10.3
12. CMake version: 3.18.0

** Bug Report **
Under a contrastive learning setup, there seems to be some numerical instability when training over 16+ GPUs. I find this issue doesn't occur within DDP.

You can reproduce with this training script:

```python
import torch
import torch.nn as nn
import torch.nn.functional as F
import numpy as np
import torch.distributed as dist
import torch.multiprocessing as mp
import horovod.torch as hvd
from torch.cuda.amp import GradScaler


class Encoder(nn.Module):
    def __init__(self):
        super(Encoder, self).__init__()
        self.linear = nn.Linear(2, 4)
        self.bn = nn.BatchNorm2d(4)
        self.relu = nn.Sigmoid()
        self.image_out = nn.Linear(4, 2)
        self.text_out = nn.Linear(4, 2)
        self.logit_scale = nn.Parameter(torch.ones([]) * np.log(1 / 0.07))

    def forward(self, x):
        out = self.relu(self.linear(x))
        return self.image_out(out), self.text_out(out)

def main_worker(gpu, ngpus_per_node, distributed, losstype='global', gathertype='none', bntype='standard', global_bs=16):
    world_size = hvd.size() if distributed else 1
    k = global_bs // world_size

    torch.manual_seed(0)

    encoder = Encoder()
    encoder = encoder.cuda(gpu)

    if bntype == 'sync':
        encoder = torch.nn.SyncBatchNorm.convert_sync_batchnorm(encoder)


    optimizer = torch.optim.AdamW(
        [p for p in encoder.parameters() if p.requires_grad],
        lr=1e-4,
    )

    if distributed:
        optimizer = hvd.DistributedOptimizer(optimizer, named_parameters=encoder.named_parameters())

        hvd.broadcast_parameters(encoder.state_dict(), root_rank=0)
    
    scaler = GradScaler()
    #should I broadcast optimizer state?

    for batch in range(11):

        if (batch % 10) == 0:
            if hvd.rank() == 0:
                print('batch', batch)
        

        optimizer.zero_grad()

        with torch.cuda.amp.autocast():
            np.random.seed(batch)

            x = torch.from_numpy(np.random.randn(global_bs, 2)).float()
            if distributed:
                x = x[k*hvd.rank():k*hvd.rank() + k]
            x = x.to(gpu)

            f_T_pergpu, f_I_pergpu = encoder(x)

            if gathertype == 'dist':
                f_T = hvd.allgather(f_T_pergpu)
                f_I = hvd.allgather(f_I_pergpu)
            elif gathertype == 'none':
                f_T = f_T_pergpu
                f_I = f_I_pergpu
            else:
                print('Error. Gathertype not found.')
                exit()

            # if (not distributed) or gpu == 0:
            #     print(f_T)
            #logit_scale = encoder.module.logit_scale if distributed else encoder.logit_scale
            logit_scale = encoder.logit_scale

            if losstype == 'global':
                logits = logit_scale * f_T @ f_I.t()
                target = torch.arange(global_bs).long().to(gpu)
                loss = F.cross_entropy(logits, target)
            elif losstype == 'local':
                assert distributed
                logits = logit_scale * f_T[k*gpu:k*gpu+k] @ f_I.t()
                target = torch.arange(k*gpu, k*gpu+k).long().to(gpu)
                loss = F.cross_entropy(logits, target)
            else:
                print('Error. Losstype not found.')
                exit()

            loss = loss * world_size
            scaler.scale(loss).backward()
            if distributed:
                optimizer.synchronize()
                #scaler.unscale_(optimizer)
                with optimizer.skip_synchronize():
                    scaler.step(optimizer)
            else:
                scaler.step(optimizer)
        
        scaler.update()
        logit_scale.grad.detach().mul_(1./world_size)

        if (batch % 10) == 0:
            if hvd.rank() == 0:
                print('loss', loss)
                print('logit_scale_grad', logit_scale.grad)
                print('weight grad', encoder.linear.weight.grad)


def main():

    # print('\n')
    # print('=' * 16)
    # print('\nDistributed global (standard bn)...\n')
    #
    # ngpus_per_node = torch.cuda.device_count()
    # print(ngpus_per_node)
    # mp.spawn(main_worker, nprocs=ngpus_per_node, args=(ngpus_per_node, True, 'global', 'dist', 'standard'))
    hvd.init()
    if torch.cuda.is_available():
        torch.cuda.set_device(hvd.local_rank())

    if hvd.rank() == 0:
        print('\n')
        print('=' * 16)
        print('\nDistributed global...\n')

    ngpus_per_node = torch.cuda.device_count()
    #mp.spawn(main_worker, nprocs=ngpus_per_node, args=(ngpus_per_node, True, 'global', 'dist', 'sync'))

    main_worker(hvd.local_rank(), ngpus_per_node, True, 'global', 'dist')

    if hvd.rank() == 0:
        print('\n')
        print('=' * 16)
        print('\nLocal...\n')

    if hvd.rank() == 0:
        main_worker(hvd.local_rank(), 1, False, losstype='global', gathertype='none')


if __name__ == "__main__":
    main()
```

The job was started with this slurm script:

```bash
#!/bin/bash -x
#SBATCH --account=cstdl
#SBATCH --nodes=4
#SBATCH --gres=gpu:4
#SBATCH --ntasks-per-node=4
#SBATCH --cpus-per-task=24
#SBATCH --time=00:10:00
#SBATCH --partition=develbooster
#SBATCH --job-name=hvd_debugging

eval "$(/p/project/ccstdl/gordon2/miniconda3/bin/conda shell.bash hook)" # init conda
conda activate open_clip

# load low-level libraries
ml purge
ml use $OTHERSTAGES
ml Stages/2020
ml GCC/10.3.0
ml OpenMPI
#ml CUDA/11.0
#ml cuDNN/8.0.2.39-CUDA-11.3
ml NCCL/2.8.3-1-CUDA-11.3
export NCCL_DEBUG=INFO
export CUDA_VISIBLE_DEVICES=0,1,2,3

cd /p/scratch/ccstdl/gordon2/bug_analysis/test_case
export PYTHONPATH="$PYTHONPATH:$PWD/src"
srun python -u test_hvd.py
```

I also tried this using an all_gather and detaching gradients but to no avail. Happy to provide any other information needed!
